invoices/invoiceregistry.go

```diff
@@ -692,8 +717,28 @@ func (i *InvoiceRegistry) processKeySend(ctx invoiceUpdateCtx) error {
 	_, err = i.AddInvoice(invoice, ctx.hash)
 	if err != nil && err != channeldb.ErrDuplicateInvoice {
 		return err
+	} else if err == channeldb.ErrDuplicateInvoice {
+		dbInvoice, err := i.LookupInvoice(ctx.hash)
+		if err != nil {
+			return err
+		}
+
+		// Extract the payment address, which was generated at random on
+		// the initial pass.
+		payAddr = dbInvoice.Terms.PaymentAddr
+
+		// If the HTLC got added to the invoice, we'll also extract the
+		// set id we generated previously.
+		htlc, ok := dbInvoice.Htlcs[ctx.circuitKey]
+		if ok && htlc.AMP != nil {
+			setID = htlc.AMP.SetID()
+		}
 	}
 
+	ctx.mpp = record.NewMPP(amt, payAddr)
+	ctx.amp = record.NewAMP(preimage, setID, 0)
```

Comment:
Alternatively, could it work to mutate the payload instead of the context?

Reply:
IMO it is preferable to mutate a struct which is owned by the invoice registry than one which provided externally.

Reply:
If you don't want to mutate, it could be a copy? I think that it is cleaner that have an independent 'pre-processing' step for legacy keysends that augments the payload and from there on all logic will be the same.

Reply:
I don't mind the mutation in this case, the flow is linear so and the possible mutations are limited to three LOC. I'm not sure the extra effort in extending the `Payload` interface with mutations methods (that must be reimplemented on all structs/mocks) + deep adding deep copying is that much better than what is already implemented. There are much bigger issues in the critical path IMO

Reply:
In general I don't think that bigger issues on the critical path are reason to leave aside smaller issues. But I agree that time should be spent in proportion. 

Another option could be to make a `Payload` wrapper. It is mostly a pass-through except for the legacy keysend modification.

But non-blocking.

Reply:
I agree we shouldn't ignore small issues, I just don't think we'd actually be solving a real issue in jumping through these hoops right now. The payload wrapper is a good suggestion though, I think we should consider it (hopefully not) if we find ourselves wanting to add more points of mutation, but for now it's a premature nice-to-have IMO